drake/solvers/Optimization.h

```diff
@@ -108,46 +116,123 @@ namespace Drake {
+            : Constraint(vars,Vector1d::Constant(lb),Vector1d::Constant(ub)), Q(Q), b(b) {}
+    virtual ~QuadraticConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=.5*x.transpose()*Q*x + b.transpose()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = .5*x.transpose()*Q.cast<TaylorVarXd>()*x + b.cast<TaylorVarXd>().transpose()*x; };
+
+  private:
+    Eigen::MatrixXd Q;
+    Eigen::VectorXd b;
+  };
+
+  class FunctionConstraint : public Constraint {
+  public:
+    FunctionConstraint(const VariableList& vars, const std::shared_ptr<DifferentiableFunction>& func, size_t num_constraints) : Constraint(vars,num_constraints), func(func) {}
+
+    template <typename DerivedLB, typename DerivedUB>
+    FunctionConstraint(const VariableList& vars, const std::shared_ptr<DifferentiableFunction>& func, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : Constraint(vars,lb,ub), func(func) {}
+    virtual ~FunctionConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override {
+      func->eval(x,y);
+    }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override {
+      func->eval(x,y);
+    }
+
+  private:
+    const std::shared_ptr<DifferentiableFunction> func;
   };
 
-  // DifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  // todo: consider implementing DifferentiableConstraint, TwiceDifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  /** LinearConstraint
+   * @brief Implements a constraint of the form @f -lb <= Ax <= ub @f
+   */
+  class LinearConstraint : public Constraint {
+  public:
+    LinearConstraint(const VariableList& vars, size_t num_constraints) : Constraint(vars, num_constraints) {}
+    template <typename DerivedLB, typename DerivedUB>
+    LinearConstraint(const VariableList& vars, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : Constraint(vars,lb,ub) {};
+    virtual ~LinearConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=getMatrix()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = getMatrix().cast<TaylorVarXd>()*x; };
+
+    virtual Eigen::SparseMatrix<double> getSparseMatrix() const { return getMatrix().sparseView(); };
+    virtual Eigen::MatrixXd getMatrix() const = 0;
+  };
+
+  class LinearConstraintImpl : public LinearConstraint {
```

Comment:
To support people (soon to happen in the dynamics class) implementing a linear constraint that evaluated via computation (e.g. of H,C,B) instead of writing to A.  Perhaps I will zap it for now, and wait til I make sure we need it.